core/dive.cpp

```diff
@@ -367,91 +367,117 @@ static bool has_unknown_used_cylinders(const struct dive &dive, const struct div
 	return num > 0;
 }
 
-void per_cylinder_mean_depth(const struct dive *dive, struct divecomputer *dc, int *mean, int *duration)
-{
-	int num_used_cylinders;
-
-	if (dive->cylinders.empty())
-		return;
-
-	for (size_t i = 0; i < dive->cylinders.size(); i++)
-		mean[i] = duration[i] = 0;
-	if (!dc)
-		return;
-
-	/*
-	 * There is no point in doing per-cylinder information
-	 * if we don't actually know about the usage of all the
-	 * used cylinders.
-	 */
-	auto used_cylinders = std::make_unique<bool[]>(dive->cylinders.size());
-	num_used_cylinders = get_cylinder_used(dive, used_cylinders.get());
-	if (has_unknown_used_cylinders(*dive, dc, used_cylinders.get(), num_used_cylinders)) {
-		/*
-		 * If we had more than one used cylinder, but
-		 * do not know usage of them, we simply cannot
-		 * account mean depth to them.
-		 */
-		if (num_used_cylinders > 1)
-			return;
-
-		/*
-		 * For a single cylinder, use the overall mean
-		 * and duration
-		 */
-		for (size_t i = 0; i < dive->cylinders.size(); i++) {
-			if (used_cylinders[i]) {
-				mean[i] = dc->meandepth.mm;
-				duration[i] = dc->duration.seconds;
-			}
-		}
-
-		return;
-	}
-	if (dc->samples.empty())
-		fake_dc(dc);
-
-	gasmix_loop loop(*dive, *dc);
-	std::vector<int> depthtime(dive->cylinders.size(), 0);
-	int lasttime = 0;
-	int lastdepth = 0;
-	int last_cylinder_index = -1;
-	std::pair<int, int> gaschange_event;
-	for (auto it = dc->samples.begin(); it != dc->samples.end(); ++it) {
-		int32_t time = it->time.seconds;
-		int depth = it->depth.mm;
-
-		/* Make sure to move the event past 'lasttime' */
-		gaschange_event = loop.cylinder_index_at(lasttime);
-
-		/* Do we need to fake a midway sample? */
-		if (last_cylinder_index >= 0 && last_cylinder_index != gaschange_event.first) {
-			int newdepth = interpolate(lastdepth, depth, gaschange_event.second - lasttime, time - lasttime);
-			if (newdepth > SURFACE_THRESHOLD || lastdepth > SURFACE_THRESHOLD) {
-				duration[gaschange_event.first] += gaschange_event.second - lasttime;
-				depthtime[gaschange_event.first] += (gaschange_event.second - lasttime) * (newdepth + lastdepth) / 2;
-			}
-
-			lasttime = gaschange_event.second;
-			lastdepth = newdepth;
-		}
-
-		/* We ignore segments at the surface */
-		if (depth > SURFACE_THRESHOLD || lastdepth > SURFACE_THRESHOLD) {
-			duration[gaschange_event.first] += time - lasttime;
-			depthtime[gaschange_event.first] += (time - lasttime) * (depth + lastdepth) / 2;
-		}
-
-		lastdepth = depth;
-		lasttime = time;
-		last_cylinder_index = gaschange_event.first;
-	}
-	for (size_t i = 0; i < dive->cylinders.size(); i++) {
-		if (duration[i])
-			mean[i] = (depthtime[i] + duration[i] / 2) / duration[i];
-	}
+void per_cylinder_mean_depth(const struct dive *dive, struct divecomputer *dc, int *mean, int *duration) {
```

Comment:
If you want to refactor this, at least take `std::vector<int> &` for mean and duration.

This would make bound checking possible.

It is insane that the caller is supposed to create arrays with the proper length and the callee can't even check it, because it is a raw pointer.

I would even go as far as creating and returning both vectors in this function.